athena-snowflake/src/main/java/com/amazonaws/athena/connectors/snowflake/SnowflakeOAuthJdbcConnectionFactory.java

```diff
@@ -0,0 +1,262 @@
+        catch (Exception ex) {
+            throw new RuntimeException("Error creating Snowflake connection: " + ex.getMessage(), ex);
+        }
+        return null;
+    }
+
+    private String loadTokenFromSecretsManager(Map<String, String> oauthConfig)
+    {
+        // Check if token related fields exist
+        if (oauthConfig.containsKey(ACCESS_TOKEN)) {
+            return oauthConfig.get(ACCESS_TOKEN);
+        }
+        return null;
+    }
+
+    private void saveTokenToSecretsManager(JSONObject tokenJson, Map<String, String> oauthConfig)
+    {
+        // Update token related fields
+        tokenJson.put(FETCHED_AT, System.currentTimeMillis() / 1000);
+        oauthConfig.put(ACCESS_TOKEN, tokenJson.getString(ACCESS_TOKEN));
+        oauthConfig.put(REFRESH_TOKEN, tokenJson.getString(REFRESH_TOKEN));
+        oauthConfig.put(EXPIRES_IN, String.valueOf(tokenJson.getInt(EXPIRES_IN)));
+        oauthConfig.put(FETCHED_AT, String.valueOf(tokenJson.getLong(FETCHED_AT)));
+
+        // Save updated secret
+        secretsClient.putSecretValue(builder -> builder
+                .secretId(this.oauthSecretName)
+                .secretString(String.valueOf(new JSONObject(oauthConfig)))
+                .build());
+    }
+
+    private String fetchAccessTokenFromSecret(Map<String, String> oauthConfig) throws Exception
+    {
+        String accessToken;
+        String clientId = Validate.notNull(oauthConfig.get(SnowflakeConstants.CLIENT_ID), "Missing required property: client_id");
+        String tokenEndpoint = Validate.notNull(oauthConfig.get(SnowflakeConstants.TOKEN_URL), "Missing required property: token_url");
+        String redirectUri = Validate.notNull(oauthConfig.get(SnowflakeConstants.REDIRECT_URI), "Missing required property: redirect_uri");
+        String clientSecret = Validate.notNull(oauthConfig.get(SnowflakeConstants.CLIENT_SECRET), "Missing required property: client_secret");
+        String authCode = Validate.notNull(oauthConfig.get(SnowflakeConstants.AUTH_CODE), "Missing required property: auth_code");
+
+        accessToken = loadTokenFromSecretsManager(oauthConfig);
+
+        if (accessToken == null) {
+            LOGGER.debug("First time auth. Using authorization_code...");
+            JSONObject tokenJson = getTokenFromAuthCode(authCode, redirectUri, tokenEndpoint, clientId, clientSecret);
+            saveTokenToSecretsManager(tokenJson, oauthConfig);
+            accessToken = tokenJson.getString(ACCESS_TOKEN);
+        }
+        else {
+            long expiresIn = Long.parseLong(oauthConfig.get(EXPIRES_IN));
+            long fetchedAt = Long.parseLong(oauthConfig.getOrDefault(FETCHED_AT, String.valueOf(0L)));
+            long now = System.currentTimeMillis() / 1000;
+
+            if ((now - fetchedAt) < expiresIn - 60) {
+                LOGGER.debug("Access token still valid.");
+            }
+            else {
+                LOGGER.debug("Access token expired. Using refresh_token...");
```

Comment:
If the token expires while the query is running or when we pull data from the driver, does anything fail?

Reply:
It should not cause any failure in the middle of the request.
Token validation is performed only once at the time of establishing the JDBC connection.
As long as the connection remains alive, the driver continues to use the same session and does not revalidate the token for ongoing queries or result processing.
So, in normal conditions, if the connection remains open throughout the query execution, there should not be any error due to token expiry.

reference - https://community.snowflake.com/s/article/FAQs-Snowflake-OAuth